core/src/main/scala/org/apache/spark/deploy/history/EventLogFileCompactor.scala

```diff
@@ -0,0 +1,235 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.deploy.history
+
+import java.io.IOException
+import java.net.URI
+import java.util.ServiceLoader
+
+import scala.collection.JavaConverters._
+import scala.io.{Codec, Source}
+
+import com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.{FileStatus, FileSystem, Path}
+import org.json4s.jackson.JsonMethods.parse
+
+import org.apache.spark.SparkConf
+import org.apache.spark.internal.Logging
+import org.apache.spark.internal.config.EVENT_LOG_ROLLING_MAX_FILES_TO_RETAIN
+import org.apache.spark.scheduler._
+import org.apache.spark.util.{JsonProtocol, Utils}
+
+class EventLogFileCompactor(
+    sparkConf: SparkConf,
+    hadoopConf: Configuration,
+    fs: FileSystem) extends Logging {
+
+  private val maxFilesToRetain: Int = sparkConf.get(EVENT_LOG_ROLLING_MAX_FILES_TO_RETAIN)
+
+  // FIXME: javadoc - caller should provide event log files (either compacted or original)
+  //  sequentially if the last event log file is already a compacted file, everything
+  //  will be skipped
+  def compact(eventLogFiles: Seq[FileStatus]): Seq[FileStatus] = {
+    if (eventLogFiles.isEmpty) {
+      return Seq.empty[FileStatus]
+    }
+
+    // skip everything if the last file is already a compacted file
+    if (EventLogFileWriter.isCompacted(eventLogFiles.last.getPath)) {
+      return Seq(eventLogFiles.last)
+    }
+
+    val (filesToCompact, filesToRetain) = findFilesToCompact(eventLogFiles)
+    if (filesToCompact.isEmpty) {
+      filesToRetain
+    } else {
+      // first pass
+      val bus = new ReplayListenerBus()
+
+      val builders = ServiceLoader.load(classOf[EventFilterBuilder],
+        Utils.getContextOrSparkClassLoader).asScala
+      builders.foreach(bus.addListener)
+
+      filesToCompact.foreach { log =>
+        Utils.tryWithResource(EventLogFileReader.openEventLog(log.getPath, fs)) { in =>
+          bus.replay(in, log.getPath.getName)
+        }
+      }
+
+      // second pass
+      val rewriter = new FilteredEventLogFileRewriter(sparkConf, hadoopConf,
+        filesToCompact, fs, builders.map(_.createFilter()).toSeq)
+      rewriter.start()
+      rewriter.rewrite()
+      rewriter.stop()
+
+      // cleanup files which are replaced with new compacted file.
```

Comment:
As I see it's possible that compacted file created successfully but the original event files was unable to delete. What's the plan there?

Reply:
As I commented, the logic doesn't break even original file and compact file co-exist for same index.